Handle loadbalancing for multiple web hosts

mhenrixon opened the pull request:
# Adding Load Balancer to Kamal

This PR adds load balancing to Kamal with minimal changes. It spreads traffic across multiple app servers while keeping things simple.

## Key Design Choices

### Simple Approach
- Uses an unreleased kamal-proxy image as the load balancer
- Add a `loadbalancer` key to your proxy settings
- Works automatically when you have multiple web hosts

### Easy Integration
- Managed through existing proxy commands
- Updates automatically during deployments
- Works without configuration when multiple web hosts are detected

### Configuration Options
- Set a specific host: `loadbalancer: lb.example.com`
- Or use default: first host becomes the load balancer
- Handles port mapping (80/443) and SSL automatically

## Why We Did It This Way

We built the load balancer into the proxy system because:

1. It's familiar to existing users
2. Reuses code we already have
3. Makes sense conceptually
4. Requires minimal setup

## How It Works

- Added load balancer capability to the proxy system
- Auto-detects when load balancing should be used
- Created simple commands for managing it
- Made it work with the normal deployment process

## Configuration Example

```yaml
proxy:
  hosts:
    - web-1.example.com
    - web-2.example.com
  loadbalancer: lb.example.com  # Optional - uses first host if not specified
  ssl: true
```

To turn off load balancing when using multiple roles:

```yaml
proxy:
  hosts:
    - web-1.example.com
    - web-2.example.com
  loadbalancer: false
  ssl: true
```

## Usage

Load balancing turns on when:
1. You add a `loadbalancer` setting, OR
2. You have multiple hosts

Basic commands:

```bash
kamal proxy loadbalancer start    # Start it
kamal proxy loadbalancer stop     # Stop it
kamal proxy loadbalancer deploy   # Update config
kamal proxy loadbalancer info     # Check status
kamal proxy loadbalancer logs     # View logs
```

## Future Plans

We can add more features later:
- Better load balancing methods
- Health checks
- Custom routing

This simple approach makes load balancing available to everyone without complicated setup.

mhenrixon committed:
Handle loadbalancing for multiple web hosts

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon committed:
Revert #1656

mhenrixon committed:
Merge remote-tracking branch 'basecamp/revert-1662-revert-1656-avoid-local-registry-login-for-remote-build' into feat/loadbalancing

mhenrixon committed:
Fix initializer to accept and provide secrets

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon committed:
Merge basecamp/main into feat/loadbalancing

Resolved conflicts:
- lib/kamal/cli/proxy.rb: Combined upstream's host-specific proxy
  pattern (KAMAL.proxy(host)) with local info logging statements
- test/integration/docker/deployer/app/config/deploy.yml: Kept both
  loadbalancer: false and run: registry: registry:4443 configurations

mhenrixon committed:
Fix MINIMUM_VERSION constant reference in loadbalancer

Update reference from Proxy::Boot::MINIMUM_VERSION to
Proxy::Run::MINIMUM_VERSION after upstream refactoring moved
the constant location.

mhenrixon committed:
Add test coverage for loadbalancer feature

- Create test/commands/loadbalancer_test.rb with 20 tests covering:
  - Container lifecycle (run, start, stop, start_or_run)
  - Deploy command with targets, SSL, and multiple hosts
  - Info, version, logs commands
  - Container/image removal and directory management

- Add 11 loadbalancer CLI tests to test/cli/proxy_test.rb:
  - boot, reboot, details with loadbalancer
  - loadbalancer subcommands (info, start, stop, logs, deploy)
  - remove_container/remove_image with loadbalancer

- Create test/fixtures/deploy_with_loadbalancer.yml fixture

mhenrixon committed:
Allow loadbalancer to share server with proxy

When the loadbalancer is configured on a server that also runs a proxy
(i.e., a web server), the loadbalancer now takes over the proxy role:

- Skip booting separate proxy container on loadbalancer host
- Loadbalancer container uses "kamal-proxy" name (not "load-balancer")
- Loadbalancer uses proxy-compatible volume mounts for app deployments
- Container label matches proxy for proper cleanup

This eliminates the port conflict (both use 80/443) and reduces the
number of containers needed when loadbalancer is on a web server.

Changes:
- Add loadbalancer_on_proxy_host? method to detect shared host
- Update boot/reboot to skip proxy on loadbalancer host
- Loadbalancer container_name is dynamic based on host type
- Add proxy-compatible volume mounts when on proxy host

Tests added for shared host scenario.

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon commented:
> Also - question: this assumes that we have at least two different targets/hosts: 1) one running Kamal proxy and 2) the second one running like a replica. Nevertheless, if we are would want a "real" loadbalancer we would need at least 3 different hosts right? 1 for the loadbalancer, and 2 others to run 2 replicas

No, i have changed it so that the extra host shouldn't be needed. 

If the current host has a rails server (web ip), i combine the load balancer and proxy to be one. 

Works as long as we have the same docker image.

dmoac commented:
> > Also - question: this assumes that we have at least two different targets/hosts: 1) one running Kamal proxy and 2) the second one running like a replica. Nevertheless, if we are would want a "real" loadbalancer we would need at least 3 different hosts right? 1 for the loadbalancer, and 2 others to run 2 replicas
> 
> No, i have changed it so that the extra host shouldn't be needed.
> 
> If the current host has a rails server (web ip), i combine the load balancer and proxy to be one.
> 
> Works as long as we have the same docker image.

Yeah! It makes total sense! If that server that is running the proxy+LB is unstable then we would have a problem of trying to route the requests

mhenrixon commented:
> If that server that is running the proxy+LB is unstable then we would have a problem of trying to route the requests

Absolutely, I just wanted the to also offer the cheapest way possible to loadbalance. 

The thing is, the load balancer and kamal proxy is not the problem and if the rails server is having problems you can always separate the two. 

Since it is running docker containers you really shouldn't have any issues. 

I have used this on fairly large projects and never had any issues.

mhenrixon committed:
Merge remote-tracking branch 'basecamp/main' into feat/loadbalancing

mhenrixon commented on lib/kamal/configuration/loadbalancer.rb:
```diff
@@ -0,0 +1,34 @@
+    opts = super
+
+    opts[:host] = hosts if hosts.present?
+    opts[:tls] = proxy_config["ssl"].presence
```
Good catch. Fixed — changed `opts[:tls] = proxy_config["ssl"].presence` to `opts[:tls] = ssl? ? true : nil` to match the boolean pattern used in `Proxy#deploy_options`.

mhenrixon commented on lib/kamal/configuration/proxy.rb:
```diff
@@ -92,6 +113,13 @@ def deploy_options
+
+    if load_balancing?
+      opts.delete(:host)
+      opts.delete(:tls)
```
Valid point. Fixed — now also deleting `:"tls-certificate-path"`, `:"tls-private-key-path"`, and `:"tls-redirect"` alongside `:host` and `:tls` in the `load_balancing?` block.

mhenrixon commented on lib/kamal/configuration/role.rb:
```diff
@@ -154,8 +154,9 @@ def asset_volume_directory(version = config.version)
-    if running_proxy? && proxy.ssl? && hosts.size > 1 && !proxy.custom_ssl_certificate?
-      raise Kamal::ConfigurationError, "SSL is only supported on a single server unless you provide custom certificates, found #{hosts.size} servers for role #{name}"
+    # Skip SSL validation when a loadbalancer is present or custom certificates are provided
+    if running_proxy? && proxy.ssl? && hosts.size > 1 && !proxy.loadbalancer.present? && !proxy.custom_ssl_certificate?
```
Good catch, but it's slightly more nuanced. Using `proxy.load_balancing?` unconditionally would skip the validation for *all* roles (including non-primary roles like workers), since `effective_loadbalancer` is based on the primary role's host count. Fixed with `!(primary? && proxy.load_balancing?)` — the SSL multi-host restriction is only relaxed for the primary role when load balancing is active. Non-primary roles with multiple hosts and SSL still need custom certificates.

mhenrixon commented on lib/kamal/configuration/validator/proxy.rb:
```diff
@@ -3,7 +3,9 @@ def validate!
       super
 
-      if config["host"].blank? && config["hosts"].blank? && config["ssl"]
+      # Skip SSL host validation when a loadbalancer is present
+      # since SSL is disabled when using a loadbalancer
+      if config["host"].blank? && config["hosts"].blank? && config["ssl"] && config["loadbalancer"].blank?
         error "Must set a host to enable automatic SSL"
       end
```
Good catch. Added explicit validation that rejects `loadbalancer: true` and other non-string types. Only a hostname string or `false` (to disable) are accepted. Also added tests for this.

mhenrixon commented on lib/kamal/configuration/docs/proxy.yml:
```diff
@@ -25,6 +25,13 @@ proxy:
+  # Loadbalancer
+  #
+  # Specify a host to run the loadbalancer on. The loadbalancer will distribute requests
+  # to all web hosts. If not specified but multiple web hosts are configured, the first
+  # web host will be used as the loadbalancer host.
+  loadbalancer: lb.example.com
```
Good point. Updated the docs to document `loadbalancer: false` as a way to explicitly disable load balancing when multiple web hosts are configured.

mhenrixon commented on lib/kamal/commands/loadbalancer.rb:
```diff
@@ -0,0 +1,130 @@
+  end
+
+  def remove_container
+    docker :container, :prune, "--force", "--filter", "label=org.opencontainers.image.title=kamal-loadbalancer"
+  end
+
+  def remove_image
+    docker :image, :prune, "--all", "--force", "--filter", "label=org.opencontainers.image.title=kamal-loadbalancer"
```
Good catch — the label-based prune wouldn't work for the shared-host case. Fixed by switching to `docker container rm --force <container_name>` and `docker image rm --force <image>` which work regardless of the label used.

mhenrixon commented on lib/kamal/commands/loadbalancer.rb:
```diff
@@ -0,0 +1,130 @@
+  def version
+    pipe \
+      docker(:inspect, container_name, "--format '{{.Config.Image}}'"),
+      [ :cut, "-d:", "-f2" ]
```
Good catch. Fixed — switched from `cut -d: -f2` to `awk -F: '{print $NF}'` to correctly handle images with registry ports (e.g. `registry.example.com:5000/kamal-proxy:1.2.3`).

mhenrixon commented on lib/kamal/cli/proxy.rb:
```diff
@@ -127,6 +150,25 @@ def reboot
           run_hook "post-proxy-reboot", hosts: host_list
         end
+
+        if KAMAL.config.proxy.load_balancing?
+          lb_host = KAMAL.config.proxy.effective_loadbalancer
+          run_hook "pre-loadbalancer-reboot", hosts: lb_host
+
+          on(lb_host) do |host|
+            execute *KAMAL.auditor.record("Rebooted loadbalancer"), verbosity: :debug
+            execute *KAMAL.registry.login
+
+            info "Stopping and removing #{KAMAL.loadbalancer.container_name} on #{host}, if running..."
+            execute *KAMAL.loadbalancer.stop, raise_on_non_zero_exit: false
+            execute *KAMAL.loadbalancer.remove_container
+            execute *KAMAL.loadbalancer.ensure_apps_config_directory
+
+            execute *KAMAL.loadbalancer.run
+          end
```
This is the same root cause as the `remove_container` comment above. Fixed by switching `remove_container` to `docker container rm --force <container_name>` which uses the dynamic `container_name` (either `kamal-proxy` or `load-balancer` depending on `on_proxy_host?`), so it works correctly in both cases.

mhenrixon commented on lib/kamal/cli/build.rb:
```diff
@@ -18,7 +18,7 @@ def push
 
     ensure_docker_installed
-    login_to_registry_locally if KAMAL.builder.login_to_registry_locally?
+    login_to_registry_locally
```
Good catch — this was a merge regression. Already fixed by restoring the conditional: `login_to_registry_locally if KAMAL.builder.login_to_registry_locally?`

mhenrixon committed:
docs: drop stale CLAUDE.md (fork docs live on the dash branch)

mhenrixon committed:
Merge branch main into feat/loadbalancing

Resolutions:
- lib/kamal/cli/main.rb: adopt upstream modify(lock: true) deploy/redeploy
  structure and post-deploy hook; move loadbalancer deploy invocation into
  the new locked block after app boot
- lib/kamal/cli/proxy.rb: adopt modify(lock: true) in reboot while keeping
  loadbalancer host skipping
- lib/kamal/cli/build.rb: keep upstream setup_local_registry, keep this
  branch unconditional login_to_registry_locally for non-local registries
- test/integration/.../app/config/deploy.yml: keep loadbalancer: false,
  drop proxy run.registry removed upstream
- Gemfile.lock: taken from main
- Adapt loadbalancer info command/tests to upstream quoted docker ps filter
- Update SSL multi-server error message expectation in configuration test

mhenrixon committed:
test: expect local registry login with remote builder

The fork reverts #1664 and always logs in locally for non-local
registries; the test still asserted the upstream skip.

mhenrixon committed:
test: disable auto-loadbalancer in multi-host integration fixtures

The loadbalancer auto-activates for any primary role with >1 host.
The integration harness can't support it (vm hostnames don't resolve
from inside the inner dind network, and the LB's hardcoded 80:443
publish conflicts with the traefik fixture), so fixtures opt out like
app and app_with_roles already do.
